File: ports/php/src/Hqtui.php
<?php
declare(strict_types=1);
namespace Hqtui;

final class ZendAdapter {
    public function __call(string $method, array $arguments): mixed { return \hqtui_bridge($method, ...$arguments); }
}
final class Native {
    private static ?object $api = null;
    public static function api(): object {
        if (self::$api !== null) return self::$api;
        if (extension_loaded('hqtui_native')) {
            if (\hqtui_bridge('hqb_abi_version') !== 1) throw new \RuntimeException('Unsupported native ABI');
            return self::$api = new ZendAdapter();
        }
        if (!extension_loaded('ffi')) throw new \RuntimeException('PHP FFI extension is required for HQTUI.');
        $path = getenv('HQTUI_NATIVE_LIB') ?: __DIR__ . '/../../cpp/build-bindings/libhqtui_bindings.' . (PHP_OS_FAMILY === 'Darwin' ? 'dylib' : 'so');
        if (!is_file($path)) throw new \RuntimeException('Native library missing. Build ports/cpp with -DHQTUI_BUILD_BINDINGS=ON or set HQTUI_NATIVE_LIB.');
        $api = \FFI::cdef(<<<'C'
typedef struct hqb_scene hqb_scene;
int hqb_abi_version(void);
const char *hqb_error(void);
hqb_scene *hqb_create(int, int, const char *);
void hqb_destroy(hqb_scene *);
int hqb_set(hqb_scene *, const char *, size_t);
int hqb_resize(hqb_scene *, int, int);
int hqb_collapse(hqb_scene *, int);
const char *hqb_render(hqb_scene *, const char *);
const char *hqb_demo_frame(hqb_scene *, const char *, const char *);
int hqb_open(hqb_scene *);
int hqb_present(hqb_scene *);
const char *hqb_poll(hqb_scene *, int);
int hqb_interrupted(void);
void hqb_close(hqb_scene *);
int hqb_demo(const char *, size_t);
C, $path);
        if ($api->hqb_abi_version() !== 1) throw new \RuntimeException('Unsupported native ABI');
        return self::$api = $api;
    }
    public static function check(mixed $value): mixed {
        if ($value === 0 || $value === null || ($value instanceof \FFI\CData && \FFI::isNull($value)))
            throw new \RuntimeException(self::string(self::api()->hqb_error()));
        return $value;
    }
    public static function string(mixed $value): string { return is_string($value) ? $value : \FFI::string($value); }
    public static function demo(array $arguments): int {
        $json = json_encode(array_values($arguments), JSON_THROW_ON_ERROR);
        $status = self::api()->hqb_demo($json, strlen($json));
        if ($status < 0) throw new \RuntimeException(self::string(self::api()->hqb_error()));
        return $status;
    }
}

final class UI implements \JsonSerializable {
    private array $node;
    public function __construct(string $type = 'col', array $options = []) { $this->node = ['type'=>$type, ...$options, 'children'=>[]]; }
    public function jsonSerialize(): array { return $this->node; }
    public function add(string $type, array $options = []): self { $this->node['children'][] = ['type'=>$type, ...$options]; return $this; }
    public function group(string $type, callable $body, array $options = []): self {
        $child = new self($type, $options); $body($child); $this->node['children'][] = $child; return $this;
    }
    public function row(callable $body, array $options = []): self { return $this->group('row', $body, $options); }
    public function col(callable $body, array $options = []): self { return $this->group('col', $body, $options); }
    public function panel(string $title, callable $body, array $options = []): self { return $this->group('panel', $body, ['title'=>$title, ...$options]); }
    public function text(string $text, array $options = []): self { return $this->add('text', ['text'=>$text, ...$options]); }
    public function meter(float $value, array $options = []): self { return $this->add('meter', ['value'=>$value, ...$options]); }
    public function graph(array $values, array $options = []): self { return $this->add('graph', ['values'=>$values, ...$options]); }
    public function gauge(float $value, array $options = []): self { return $this->add('gauge', ['value'=>$value, ...$options]); }
    public function table(array $columns, array $rows, array $options = []): self { return $this->add('table', ['columns'=>$columns, 'rows'=>$rows, ...$options]); }
    public function keys(array $rows, array $options = []): self { return $this->add('keys', ['rows'=>$rows, ...$options]); }
    public function log(array $entries, array $options = []): self { return $this->add('log', ['entries'=>$entries, ...$options]); }
    public function spacer(array $options = []): self { return $this->add('spacer', $options); }
    public function divider(string $text = ''): self { return $this->add('divider', ['text'=>$text]); }
    public function badge(string $text, array $o = []): self { return $this->add('badge', ['text'=>$text, ...$o]); }
    public function progress(float $value, array $o = []): self { return $this->add('progress', ['value'=>$value, ...$o]); }
    public function sparkline(array $values, array $o = []): self { return $this->add('sparkline', ['values'=>$values, ...$o]); }
    public function heatbar(float $value, array $o = []): self { return $this->add('heatbar', ['value'=>$value, ...$o]); }
    public function columns(array $values, array $o = []): self { return $this->add('columns', ['values'=>$values, ...$o]); }
    public function donut(array $segments, array $o = []): self { return $this->add('donut', ['segments'=>$segments, ...$o]); }
    public function list(array $items, array $o = []): self { return $this->add('list', ['items'=>$items, ...$o]); }
    public function tree(array $nodes, array $o = []): self { return $this->add('tree', ['nodes'=>$nodes, ...$o]); }
    public function button(string $label, array $o = []): self { return $this->add('button', ['label'=>$label, ...$o]); }
    public function checkbox(string $label, array $o = []): self { return $this->add('checkbox', ['label'=>$label, ...$o]); }
    public function select(string $value, array $o = []): self { return $this->add('select', ['value'=>$value, ...$o]); }
    public function input(string $value, array $o = []): self { return $this->add('input', ['value'=>$value, ...$o]); }
    public function tabs(array $tabs, array $o = []): self { return $this->add('tabs', ['tabs'=>$tabs, ...$o]); }
    public function statusbar(array $items, array $o = []): self { return $this->add('statusbar', ['items'=>$items, ...$o]); }
    public function label(string $text, array $o = []): self { return $this->add('label', ['text'=>$text, ...$o]); }
    public function heading(string $text, array $o = []): self { return $this->add('heading', ['text'=>$text, ...$o]); }
    public function meters(array $items, array $o = []): self { return $this->add('meters', ['items'=>$items, ...$o]); }
    // Overlays take the whole screen, so they carry no size and are drawn last.
    public function modal(array $o = []): self { return $this->add('modal', $o); }
    public function commandPalette(array $o = []): self { return $this->add('commandpalette', $o); }
    public function tooltip(string $text, int $x, int $y, array $o = []): self { return $this->add('tooltip', ['text'=>$text, 'x'=>$x, 'y'=>$y, ...$o]); }
}

final class Scene {
    private mixed $handle = null;
    private object $native;
    public function __construct(int $width=80, int $height=24, string $theme='dark') {
        $this->native = Native::api();
        $this->handle = Native::check($this->native->hqb_create($width,$height,$theme));
    }
    private function __clone() {}
    public function __serialize(): array { throw new \RuntimeException('Scenes cannot be serialized'); }
    private function handle(): mixed {
        if ($this->handle === null) throw new \RuntimeException('Scene is closed');
        return $this->handle;
    }
    public function set(UI|array $ui): self {
        $json = json_encode($ui, JSON_THROW_ON_ERROR);
        Native::check($this->native->hqb_set($this->handle(), $json, strlen($json))); return $this;
    }
    /** Merge the borders of adjacent panels into shared lines. */
    public function collapse(bool $enabled = true): self { Native::check($this->native->hqb_collapse($this->handle(),$enabled?1:0)); return $this; }
    public function resize(int $width, int $height): self { Native::check($this->native->hqb_resize($this->handle(),$width,$height)); return $this; }
    public function render(string $format='text'): string { return Native::string(Native::check($this->native->hqb_render($this->handle(),$format))); }
    public function demoFrame(string $screen, string $format='text'): string { return Native::string(Native::check($this->native->hqb_demo_frame($this->handle(),$screen,$format))); }
    public function present(): void { Native::check($this->native->hqb_present($this->handle())); }
    public function poll(int $timeoutMs=33): string { return Native::string(Native::check($this->native->hqb_poll($this->handle(),$timeoutMs))); }
    public function interrupted(): bool { return $this->native->hqb_interrupted() !== 0; }
    public function withTerminal(callable $body): mixed {
        Native::check($this->native->hqb_open($this->handle()));
        try { return $body($this); }
        finally { if ($this->handle !== null) $this->native->hqb_close($this->handle); }
    }
    public function close(): void {
        if ($this->handle !== null) { $this->native->hqb_destroy($this->handle); $this->handle=null; }
    }
    public function __destruct() { $this->close(); }
}

File: ports/ruby/lib/hqtui.rb
# frozen_string_literal: true
require 'json'
require 'fiddle/import'

module Hqtui
  VERSION = '0.2.0'
  class Error < StandardError; end
  def self.native
    @native ||= Module.new do
      extend Fiddle::Importer
      path = ENV['HQTUI_NATIVE_LIB'] || File.expand_path('../../cpp/build-bindings/libhqtui_bindings.' + (RUBY_PLATFORM.include?('darwin') ? 'dylib' : 'so'), __dir__)
      raise Error, 'Native library missing. Build ports/cpp with -DHQTUI_BUILD_BINDINGS=ON or set HQTUI_NATIVE_LIB.' unless File.file?(path)
      dlload path
      extern 'int hqb_abi_version()'
      extern 'const char* hqb_error()'
      extern 'void* hqb_create(int, int, const char*)'
      extern 'void hqb_destroy(void*)'
      extern 'int hqb_set(void*, const char*, size_t)'
      extern 'int hqb_resize(void*, int, int)'
      extern 'int hqb_collapse(void*, int)'
      extern 'const char* hqb_render(void*, const char*)'
      extern 'const char* hqb_demo_frame(void*, const char*, const char*)'
      extern 'int hqb_open(void*)'
      extern 'int hqb_present(void*)'
      extern 'const char* hqb_poll(void*, int)'
      extern 'int hqb_interrupted()'
      extern 'void hqb_close(void*)'
      extern 'int hqb_demo(const char*, size_t)'
      raise Error, 'Unsupported native ABI' unless hqb_abi_version == 1
    end
  end
  def self.check(value)
    raise Error, native.hqb_error.to_s if value == 0 || value.nil? || (value.respond_to?(:null?) && value.null?)
    value
  end
  def self.demo(arguments = ARGV)
    encoded = JSON.generate(arguments)
    status = native.hqb_demo(encoded, encoded.bytesize)
    raise Error, native.hqb_error.to_s if status < 0
    status
  end
  class UI
    def initialize(type = 'col', **options)
      @node = { type: type, **options, children: [] }
    end
    def to_h = @node
    def add(type, **options)
      @node[:children] << { type: type, **options }; self
    end
    def group(type, **options)
      child = UI.new(type, **options)
      yield child if block_given?
      @node[:children] << child.to_h; self
    end
    def row(**options, &block) = group('row', **options, &block)
    def col(**options, &block) = group('col', **options, &block)
    def panel(title, **options, &block) = group('panel', title: title, **options, &block)
    def text(text, **options) = add('text', text: text, **options)
    def meter(value, **options) = add('meter', value: value, **options)
    def graph(values, **options) = add('graph', values: values, **options)
    def gauge(value, **options) = add('gauge', value: value, **options)
    def table(columns, rows, **options) = add('table', columns: columns, rows: rows, **options)
    def keys(rows, **options) = add('keys', rows: rows, **options)
    def log(entries, **options) = add('log', entries: entries, **options)
    def spacer(**options) = add('spacer', **options)
    def divider(text = '') = add('divider', text: text)
    def badge(text, **options) = add('badge', text: text, **options)
    def progress(value, **options) = add('progress', value: value, **options)
    def sparkline(values, **options) = add('sparkline', values: values, **options)
    def heatbar(value, **options) = add('heatbar', value: value, **options)
    def columns(values, **options) = add('columns', values: values, **options)
    def donut(segments, **options) = add('donut', segments: segments, **options)
    def list(items, **options) = add('list', items: items, **options)
    def scrollbar(total, **options) = add('scrollbar', total: total, **options)
    def chart(series, **options) = add('chart', series: series, **options)
    def calendar(year, month, **options) = add('calendar', year: year, month: month, **options)
    def tree(nodes, **options) = add('tree', nodes: nodes, **options)
    def button(label, **options) = add('button', label: label, **options)
    def checkbox(label, **options) = add('checkbox', label: label, **options)
    def select(value, **options) = add('select', value: value, **options)
    def input(value, **options) = add('input', value: value, **options)
    def tabs(tabs, **options) = add('tabs', tabs: tabs, **options)
    def statusbar(items, **options) = add('statusbar', items: items, **options)
    def label(text, **options) = add('label', text: text, **options)
    def heading(text, **options) = add('heading', text: text, **options)
    def meters(items, **options) = add('meters', items: items, **options)
    # Overlays take the whole screen, so they carry no size and are drawn last.
    def modal(**options) = add('modal', **options)
    def command_palette(**options) = add('commandpalette', **options)
    def tooltip(text, x:, y:, **options) = add('tooltip', text: text, x: x, y: y, **options)
  end
  class Scene
    def self.finalizer(native, pointer) = proc { native.hqb_destroy(pointer) }
    def initialize(width: 80, height: 24, theme: 'dark')
      @native = Hqtui.native
      @handle = Hqtui.check(@native.hqb_create(width, height, theme))
      ObjectSpace.define_finalizer(self, self.class.finalizer(@native, @handle))
    end
    def initialize_copy(*) = raise(Error, 'Scenes cannot be copied')
    def handle
      raise Error, 'Scene is closed' unless @handle
      @handle
    end
    def set(ui)
      encoded = JSON.generate(ui.respond_to?(:to_h) ? ui.to_h : ui)
      Hqtui.check(@native.hqb_set(handle, encoded, encoded.bytesize)); self
    end
    # Merge the borders of adjacent panels into shared lines.
    def collapse(enabled = true)
      Hqtui.check(@native.hqb_collapse(handle, enabled ? 1 : 0)); self
    end

    def resize(width, height)
      Hqtui.check(@native.hqb_resize(handle, width, height)); self
    end
    def render(format = 'text') = Hqtui.check(@native.hqb_render(handle, format)).to_s.force_encoding(Encoding::UTF_8)
    def demo_frame(screen, format = 'text') = Hqtui.check(@native.hqb_demo_frame(handle, screen, format)).to_s.force_encoding(Encoding::UTF_8)
    def present = Hqtui.check(@native.hqb_present(handle))
    def poll(timeout_ms = 33) = Hqtui.check(@native.hqb_poll(handle, timeout_ms)).to_s
    def interrupted? = @native.hqb_interrupted != 0
    def with_terminal
      Hqtui.check(@native.hqb_open(handle))
      begin
        yield self
      ensure
        @native.hqb_close(handle) if @handle
      end
    end
    def close
      return unless @handle
      ObjectSpace.undefine_finalizer(self)
      @native.hqb_destroy(@handle); @handle = nil
    end
  end
end
